File: scenarios/audio-tools/ui/src/audio-integration/features.ts
// features.ts — canonical capability / feature identifiers for audio-tools.
//
// Mirror of clients/go/audiotools/features.go. Consumer scenarios
// register a dependency-capability under AUDIO_TOOLS_CAPABILITY_SLUG and
// list the feature slugs they wire into their UI. The slugs are the
// wire-level identifiers carried on web-console's CapabilitiesService
// `repeated string features` surface; they must agree byte-for-byte
// across the API registry and every UI consumer.
//
// Drift safety: the slug map is keyed by the generated AudioToolsFeature
// proto enum. Adding a new value to the enum without updating the map
// triggers a runtime throw at module load (see assertCoverage below) so
// the bug is caught before any UI render path uses a stale slug.
//
// Adding a new feature: add the enum value in
// packages/proto/schemas/audio-tools/v1/common/common.proto, run
// `make generate`, then add the slug entry below AND in
// clients/go/audiotools/features.go.
//
// IMPORTANT: this file lives verbatim in audio-tools/ui (canonical) and
// is copied into each consumer scenario. Update the source here, then
// copy to web-console/ui and swarm-manager/ui.

import { AudioToolsFeature } from "@vrooli/proto-types/audio-tools/v1/common/common_pb";

/** Scenario-level capability id used when registering audio-tools as a dependency-capability. Matches the scenario slug. */
export const AUDIO_TOOLS_CAPABILITY_SLUG = "audio-tools";

const FEATURE_SLUGS: Record<AudioToolsFeature, string> = {
  [AudioToolsFeature.UNSPECIFIED]: "",
  [AudioToolsFeature.VOICE_INPUT]: "voice-input",
  [AudioToolsFeature.VOICE_STREAMING]: "voice-streaming",
  [AudioToolsFeature.VOICE_SPEAKER_VERIFICATION]: "voice-speaker-verification",
  [AudioToolsFeature.VOICE_ENROLLMENT]: "voice-enrollment",
  [AudioToolsFeature.VOICE_OUTPUT]: "voice-output",
  [AudioToolsFeature.TTS_SUMMARIZATION]: "tts-summarization",
  [AudioToolsFeature.TTS_CACHE]: "tts-cache",
  [AudioToolsFeature.TTS_PARAGRAPH_SPLIT]: "tts-paragraph-split",
  [AudioToolsFeature.AUDIO_PROVIDER_ROUTING]: "audio-provider-routing",
};

// The numeric members of the AudioToolsFeature enum. Object.values of a
// numeric enum yields both the reverse-mapped names (strings) and the
// numeric values; the type guard keeps only the latter as enum values.
const FEATURE_VALUES = Object.values(AudioToolsFeature).filter(
  (v): v is AudioToolsFeature => typeof v === "number",
);

(function assertCoverage() {
  // FEATURE_SLUGS is typed Record<AudioToolsFeature, string>, so the compiler
  // already enforces an entry per enum member; this runtime guard catches a
  // hand-edited empty slug and points at the Go mirror that must also change.
  for (const value of FEATURE_VALUES) {
    if (value === AudioToolsFeature.UNSPECIFIED) continue;
    if (!FEATURE_SLUGS[value]) {
      throw new Error(
        `audio-integration/features.ts: missing slug for AudioToolsFeature.${AudioToolsFeature[value]} (${value}). ` +
          "Add the entry to FEATURE_SLUGS and to clients/go/audiotools/features.go.",
      );
    }
  }
})();

/** Returns the wire-slug for an AudioToolsFeature enum value. Empty string for UNSPECIFIED. */
export function featureSlug(f: AudioToolsFeature): string {
  return FEATURE_SLUGS[f];
}

/** All registered feature slugs in stable (alphabetical) order. */
export function allFeatureSlugs(): string[] {
  return Object.values(FEATURE_SLUGS).filter((s) => s !== "").sort();
}

export { AudioToolsFeature };
// HOST DIFFERENCE: audio-tools maps its generated feature enum for its local API surface.

File: scenarios/audio-tools/clients/go/audiotools/features.go
// features.go — canonical capability / feature identifiers for audio-tools.
//
// Consumer scenarios (web-console, swarm-manager, ...) register a
// dependency-capability under CapabilitySlug and list the feature slugs
// they wire into their UI. The slugs are the wire-level identifiers
// carried on web-console's CapabilitiesService `repeated string features`
// surface; they must agree byte-for-byte across the API registry and
// every UI consumer.
//
// To prevent drift, callers MUST use:
//
//	audiotools.CapabilitySlug                              // "audio-tools"
//	audiotools.FeatureSlug(common_v1.AudioToolsFeature_*)  // "voice-input", ...
//	audiotools.AllFeatureSlugs()                           // every registered slug
//
// Adding a new feature: add the enum value in
// packages/proto/schemas/audio-tools/v1/common/common.proto, run
// `make generate`, then add the slug entry in `featureSlugs` below AND
// in audio-tools/ui/src/audio-integration/features.ts.

package audiotools

import (
	"fmt"
	"sort"

	common_v1 "github.com/vrooli/vrooli/packages/proto/gen/go/audio-tools/v1/common"
)

// CapabilitySlug is the scenario-level capability id used when registering
// audio-tools as a dependency capability. Matches the scenario slug.
const CapabilitySlug = "audio-tools"

// featureSlugs is the single source of truth for the AudioToolsFeature
// enum → wire-slug mapping. Keys MUST cover every value of the enum
// except _UNSPECIFIED; a missing key triggers a panic at startup via
// the init() check below.
var featureSlugs = map[common_v1.AudioToolsFeature]string{
	common_v1.AudioToolsFeature_AUDIO_TOOLS_FEATURE_VOICE_INPUT:                "voice-input",
	common_v1.AudioToolsFeature_AUDIO_TOOLS_FEATURE_VOICE_STREAMING:            "voice-streaming",
	common_v1.AudioToolsFeature_AUDIO_TOOLS_FEATURE_VOICE_SPEAKER_VERIFICATION: "voice-speaker-verification",
	common_v1.AudioToolsFeature_AUDIO_TOOLS_FEATURE_VOICE_ENROLLMENT:           "voice-enrollment",
	common_v1.AudioToolsFeature_AUDIO_TOOLS_FEATURE_VOICE_OUTPUT:               "voice-output",
	common_v1.AudioToolsFeature_AUDIO_TOOLS_FEATURE_TTS_SUMMARIZATION:          "tts-summarization",
	common_v1.AudioToolsFeature_AUDIO_TOOLS_FEATURE_TTS_CACHE:                  "tts-cache",
	common_v1.AudioToolsFeature_AUDIO_TOOLS_FEATURE_TTS_PARAGRAPH_SPLIT:        "tts-paragraph-split",
	common_v1.AudioToolsFeature_AUDIO_TOOLS_FEATURE_AUDIO_PROVIDER_ROUTING:     "audio-provider-routing",
}

func init() {
	// Fail-fast if a generated enum value lacks a slug entry. This catches
	// the case where someone adds an enum value, regenerates, and forgets
	// to update featureSlugs (or features.ts).
	for v, name := range common_v1.AudioToolsFeature_name {
		feat := common_v1.AudioToolsFeature(v)
		if feat == common_v1.AudioToolsFeature_AUDIO_TOOLS_FEATURE_UNSPECIFIED {
			continue
		}
		if _, ok := featureSlugs[feat]; !ok {
			panic(fmt.Sprintf("audiotools: missing slug for AudioToolsFeature value %s (%d)", name, v))
		}
	}
}

// FeatureSlug returns the wire-slug for an AudioToolsFeature enum value.
// Returns the empty string for _UNSPECIFIED or any value not present in
// the mapping table.
func FeatureSlug(f common_v1.AudioToolsFeature) string {
	return featureSlugs[f]
}

// AllFeatureSlugs returns every registered feature slug in stable
// (alphabetical) order. Useful for consumer scenarios that want to
// register the full audio-tools feature surface.
func AllFeatureSlugs() []string {
	out := make([]string, 0, len(featureSlugs))
	for _, s := range featureSlugs {
		out = append(out, s)
	}
	sort.Strings(out)
	return out
}
